### ISSUE TITLE
Instance destruction postponed

### ISSUE THREAD

**Issue by ohreally** [sdgathman/pymilter#56]
Hi,

I added a `__del__(self)` method to my milter class to log a message when an instance is destroyed, and I noticed that if libmilter triggers a warning, like
`milter claimed not to reply in state 7 but did anyway 4`
or
`st_optionneg[34386130944]: xxfi_negotiate returned 1 (protocol options=0x1fffff, actions=0x1ff)`
the instance is not destroyed until another instance triggers a libmilter warning.
I was wondering if this could pose a problem on a server that's busier than mine, where milter instances could 'overlap' (to me it looks like a variable reference that is not released, and can only be bound to a single instance at a time).

And I have a question that is somewhat related.
The first warning mentioned above is probably triggered by `return Milter.CONTINUE` in combination with the `@Milter.noreply` decorator. Is there a way to find out whether `noreply` negotiation succeeded, so that I can make the CONTINUE conditional?

Thanks in advance,
  Rob

**Comment by sdgathman** [comment]
You are correct.  However, libmilter controls when a connection (represented on our end by an instance) is destroyed.  This is triggered by the close_callback.  A mitigation might be to destroy the python part early.  The milter_wrap_close() in miltermodule.c does nothing if no python part exists.  In fact, libmilter can close a connection before it even does the connect_callback.  

The tricky part is knowing when no other callbacks will be made - or at least handling unexpected callbacks cleanly.  Currently,  it *looks* like callbacks return TEMPFAIL when there is no python object on a callback other than CONNECT or CLOSE.

**Comment by sdgathman** [comment]
> The first warning mentioned above is probably triggered by return Milter.CONTINUE in combination with the @Milter.noreply decorator. Is there a way to find out whether noreply negotiation succeeded, so that I can make the CONTINUE conditional?

The docs for @noreply decorator explain that you should return CONTINUE in case the negotiation fails.  The conditional return value is in the decorator.  It returns NOREPLY on success, CONTINUE otherwise, and throws a python exception if the decorated function returns anything other than CONTINUE.

**Comment by ohreally** [comment]
That's what I thought, but I got that exception anyway. However, a few days ago I changed some small (unrelated, I thought) things in my code, and the exception is gone.
So I guess the second problem in my original post solved itself, and I'm sorry to have wasted your time.

**Comment by sdgathman** [comment]
Not a waste at all.  Good code review.  This could be a problem.  I don't have any good ideas for *when* to destroy the python object early.  So this will have to go on the back burner.
